### PR TITLE
fix: ignore zero-action samples in group reward baselines

### PR DESCRIPTION
**Pull request by haoyang9804**
## Summary

This fixes a quiet reward/advantage bug in `RemoteExperienceMaker.compute_advantages_and_returns`.

For group-style estimators such as `dr_grpo`, OpenRLHF centers rewards within `n_samples_per_prompt` groups. A rollout sibling can have `action_mask.sum() == 0` after response truncation/filtering while still carrying a scalar reward. Before this change, that zero-action sibling participated in the group reward mean and shifted the advantages of valid siblings.

The fix builds a per-sample `has_actions` mask from `action_mask`, excludes zero-action samples from group reward means/stds/baselines, and assigns zero shaped reward to zero-action samples.

## Concrete Failure Shape

The minimal failing group is two rollout siblings from the same prompt:

```text
n_samples_per_prompt: 2
advantage_estimator: dr_grpo

sample index:      [0, 1]
raw reward:        [100.0, 1.0]
action_mask.sum(): [0, 2]

zero-action action_mask:
  [[0, 0, 0, 0, 0]]

valid action_mask:
  [[0, 0, 0, 0, 0, 1, 1]]
```

The first sibling has no trainable action tokens after response processing, but
it still carries reward `100.0`. Before this fix, `dr_grpo` reward centering
used both siblings:

```text
wrong group mean:              (100.0 + 1.0) / 2 = 50.5
wrong valid centered reward:   1.0 - 50.5 = -49.5
wrong valid token advantages:  [-49.5, -49.5, ...]
```

That silently turns the accepted sibling's policy-gradient signal negative even
though the rejected/empty sibling has no trainable tokens.

After this fix, group reward statistics only use siblings with at least one
trainable action token:

```text
fixed group mean:              1.0
fixed valid centered reward:   1.0 - 1.0 = 0.0
fixed zero-action reward:      0.0
fixed valid token advantages:  [0.0, 0.0, ...]
```

## Reproduction Recipe

```json
{
  "bug_id": "BUG-FC8EB1E7D0C6",
  "target": "openrlhf",
  "boundaries": [
    "vllm.LLM.generate",
    "openrlhf.trainer.ppo_utils.samples_generator.SamplesGenerator._process_response_into_experience",
    "openrlhf.trainer.ppo_utils.experience_maker.RemoteExperienceMaker.compute_advantages_and_returns"
  ],
  "model": "Qwen/Qwen2.5-0.5B-Instruct",
  "scenario": {
    "n_samples_per_prompt": 2,
    "advantage_estimator": "dr_grpo",
    "zero_action_sibling": {
      "reward": 100.0,
      "action_mask_sum": 0
    },
    "valid_sibling": {
      "reward": 1.0,
      "action_mask_sum": 2
    }
  },
  "expected_unpatched": {
    "valid_action_advantages": "negative because zero-action reward enters the group mean"
  },
  "expected_fixed": {
    "zero_action_sample_excluded_from_group_baseline": true,
    "valid_action_advantages_not_shifted_by_zero_action_reward": true
  }
}
```

## Portable Runner

```bash
#!/usr/bin/env bash
set -euo pipefail

REPO="${REPO:-$(pwd)}"
MODEL_ID="${MODEL_ID:-Qwen/Qwen2.5-0.5B-Instruct}"
export PYTHONPATH="${REPO}${PYTHONPATH:+:${PYTHONPATH}}"
export VLLM_USE_V1="${VLLM_USE_V1:-0}"

python3 - <<'PY'
import json
import os
from types import SimpleNamespace

import torch
from huggingface_hub import snapshot_download
from transformers import AutoTokenizer
from vllm import LLM, SamplingParams

from openrlhf.trainer.ppo_utils.experience_maker import RemoteExperienceMaker
from openrlhf.trainer.ppo_utils.samples_generator import SamplesGenerator


def make_args():
    return SimpleNamespace(
        rollout=SimpleNamespace(n_samples_per_prompt=2, max_new_tokens=16),
        data=SimpleNamespace(max_len=16),
        reward=SimpleNamespace(clip_range=None, overlong_buffer_len=None, stop_properly_penalty_coef=None),
        algo=SimpleNamespace(
            advantage=SimpleNamespace(estimator="dr_grpo", gamma=1.0, lambd=1.0, no_std_norm=True)
        ),
    )


def make_maker(args):
    maker = object.__new__(RemoteExperienceMaker)
    maker.strategy = SimpleNamespace(args=args)
    maker.args = args
    maker.advantage_estimator = args.algo.advantage.estimator
    maker.kl_ctl = SimpleNamespace(value=0.0)
    return maker


def process_response(response, max_len):
    generator = object.__new__(SamplesGenerator)
    return SamplesGenerator._process_response_into_experience(generator, response, max_len=max_len)


model_id = os.environ["MODEL_ID"]
model_path = snapshot_download(model_id, local_files_only=True)
tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
prompt_ids = tokenizer(text="Answer with one short token:", add_special_tokens=False, return_tensors="pt")[
    "input_ids"
][0].tolist()

llm = LLM(
    model=model_path,
    trust_remote_code=True,
    skip_tokenizer_init=True,
    tensor_parallel_size=1,
    gpu_memory_utilization=0.20,
    max_model_len=256,
    enforce_eager=True,
)
outputs = llm.generate(
    [{"prompt_token_ids": prompt_ids}],
    SamplingParams(max_tokens=2, min_tokens=0, temperature=0.0, logprobs=1),
)
action_ids = list(outputs[0].outputs[0].token_ids)
if not action_ids:
    raise RuntimeError("vLLM produced no tokens; cannot construct the zero-action truncation scenario")

observation_tokens = prompt_ids + action_ids
response = {
    "prompt": "Answer with one short token:",
    "label": "label",
    "images": None,
    "mm_train_inputs": None,
    "observation_tokens": observation_tokens,
    "action_ranges": [(len(prompt_ids), len(observation_tokens))],
    "rollout_log_probs": None,
    "truncated": False,
    "reward": 100.0,
    "scores": 1.0,
    "extra_logs": {},
}

zero_action = process_response(response, max_len=len(prompt_ids))
valid_response = dict(response)
valid_response["reward"] = 1.0
valid_action = process_response(valid_response, max_len=len(observation_tokens))

for idx, exp in enumerate([zero_action, valid_action]):
    exp.index = [idx]
    exp.kl = torch.zeros_like(exp.action_mask, dtype=torch.float32)
    exp.values = torch.zeros_like(exp.action_mask, dtype=torch.float32)

zero_after, valid_after = make_maker(make_args()).compute_advantages_and_returns([zero_action, valid_action])

print(json.dumps({
    "real_vllm_generated_token_count": len(action_ids),
    "zero_action_mask_sum": int(zero_after.action_mask.sum().item()),
    "valid_action_mask_sum": int(valid_after.action_mask.sum().item()),
    "valid_action_advantages": valid_after.advantages.detach().cpu().tolist(),
    "valid_action_info_return": valid_after.info["return"].detach().cpu().tolist(),
}, indent=2))
PY
```

## Observed Output

Unpatched OpenRLHF `main@c3188af37cec984614aaa38906e71fa2fc57b079`:

```json
{
  "zero_action_mask_sum": 0,
  "valid_action_mask_sum": 2,
  "valid_action_advantages": [[-49.5, -49.5, -49.5, -49.5, -49.5, -49.5, -49.5]],
  "valid_action_info_return": [-49.5],
  "candidate_bug_reproduced": true
}
```

Fixed branch `94d06f391e534e8abe68a6a343fddf19ab5c4a26`:

```json
{
  "zero_action_mask_sum": 0,
  "valid_action_mask_sum": 2,
  "valid_action_advantages": [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]],
  "valid_action_info_return": [0.0],
  "candidate_bug_reproduced": false
}
```

The fixed validation uses real vLLM token-id generation and the real OpenRLHF response-processing and advantage-computation boundaries.

## Duplicate Search

OpenRLHF has no PR template. I followed `CONTRIBUTING.md`, which asks contributors to use pre-commit hooks.

Searched OpenRLHF upstream PRs/issues for exact and related reports:

- `repo:OpenRLHF/OpenRLHF is:pr dr_grpo zero action reward group baseline`: 0 results
- `repo:OpenRLHF/OpenRLHF is:pr compute_advantages_and_returns dr_grpo action_mask reward`: 0 results
- `repo:OpenRLHF/OpenRLHF is:issue dr_grpo zero action reward group baseline`: 0 results

Related but not duplicate:

- `OpenRLHF#924`: related `dr_grpo` discussion, not a zero-action reward contamination fix.
- `OpenRLHF#1115`: broad action-mask/search hit, but it changes distributed backend plumbing, not this reward-centering path.

## Tests

```bash
python3 -m pytest tests/test_experience_maker.py tests/test_loss_aggregation.py -q
python3 -m ruff check openrlhf/trainer/ppo_utils/experience_maker.py tests/test_experience_maker.py
pre-commit run --files openrlhf/trainer/ppo_utils/experience_maker.py tests/test_experience_maker.py
```

Results:

- `tests/test_experience_maker.py tests/test_loss_aggregation.py`: 8 passed
- `ruff`: passed
- `pre-commit`: passed

### CODE REVIEW

**Review by Excelius-Wang — COMMENTED** [review]
The action mask needs to follow the same index reordering as the rewards; I checked that part with experiences supplied out of prompt order.

For dr_grpo, RLOO and group normalization, valid siblings now get the expected baselines without the zero-action sibling's reward. I also checked all-valid groups against the existing formulas and all-empty groups for finite zero advantages. The two regressions and these additional checks pass. This was CPU validation of the production advantage code with unused Ray imports isolated, not a distributed rollout test.
